### PR TITLE
Modify UI for qubit tapering functions

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
This PR renames the qubit tapering functions, removes a wrapper function used for generating symmetries and PauliX ops and makes the general-purpose user-facing functions  `taper`, `symmetry_generators`, `paulix_ops` accessible with qml.functiion_name.

**Description of the Change:**
The following functions are renamed:
```
get_generators  ->  symmetry_generators
generate_paulis -> paulix_ops
transform_hamiltonian -> taper
transform_hf -> taper_hf

```
The functions  `taper`, `symmetry_generators`, `paulix_ops` are added to `pennylane.__init__`.

**Benefits:**
The UI is more intuitive and there is flexibility in bringing some general-purpose functions out of the chemistry modules without changing the UI.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/__init__.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
 import pennylane.templates
 import pennylane.hf
 import pennylane.qchem
+from pennylane.hf import taper, symmetry_generators, paulix_ops
```
Comment (soranjh) [inline pennylane/__init__.py:36]: This is done to make these 3 general user-facing functions available with `qml.name`. This guarantees that the tapering UI will not change when we bring these functions dipper inside pennylane (yet TBD where).

File: pennylane/hf/tapering.py [hunk 2]
```diff
@@ -217,84 +231,26 @@ def generate_paulis(generators, num_qubits):
...
                 bmatrest[:, col], np.zeros(bmat.shape[0] - 1, dtype=int)
             ):
-                paulix_ops.append(qml.PauliX(col))
+                paulixops.append(qml.PauliX(col))
                 break
 
-    return paulix_ops
-
-
-def generate_symmetries(qubit_op, num_qubits):
```
Comment (soranjh) [inline pennylane/hf/tapering.py:242]: This wrapper function was accepting the untapered Hamiltonian and making few calls to return the symmetry generators and paulix ops. The user now directly computes generators and paulix ops, so no need to have thus wrapper function.

File: pennylane/hf/tapering.py [hunk 3]
```diff
@@ -526,9 +482,9 @@ def transform_hf(generators, paulix_ops, paulix_sector, num_electrons, num_wires
     >>> mol = hf.Molecule(symbols, geometry, charge=1)
     >>> H = hf.generate_hamiltonian(mol)(geometry)
     >>> n_qubits, n_elec = len(H.wires), mol.n_electrons
-    >>> generators, paulix_ops = hf.generate_symmetries(H, n_qubits)
+    >>> generators, paulixops = hf.generate_symmetries(H, n_qubits)
     >>> paulix_sector = hf.optimal_sector(H, generators, n_elec)
-    >>> hf.transform_hf(generators, paulix_ops, paulix_sector,
+    >>> hf.transform_hf(generators, paulixops, paulix_sector,
```
Comment (obliviateandsurrender) [inline pennylane/hf/tapering.py:487]:
```suggestion
    >>> hf.taper_hf(generators, paulixops, paulix_sector,
```

File: pennylane/hf/tapering.py [hunk 4]
```diff
@@ -395,15 +351,15 @@ def transform_hamiltonian(h, generators, paulix_ops, paulix_sector):
     >>> geometry = np.array([[0.0, 0.0, -0.69440367], [0.0, 0.0, 0.69440367]])
     >>> mol = qml.hf.Molecule(symbols, geometry)
     >>> H = qml.hf.generate_hamiltonian(mol)(geometry)
-    >>> generators, paulix_ops = qml.hf.generate_symmetries(H, len(H.wires))
+    >>> generators, paulixops = qml.hf.generate_symmetries(H, len(H.wires))
     >>> paulix_sector = [1, -1, -1]
-    >>> H_tapered = transform_hamiltonian(H, generators, paulix_ops, paulix_sector)
+    >>> H_tapered = transform_hamiltonian(H, generators, paulixops, paulix_sector)
```
Comment (obliviateandsurrender) [inline pennylane/hf/tapering.py:356]:
```suggestion
    >>> H_tapered = taper(H, generators, paulixops, paulix_sector)
```

File: pennylane/hf/tapering.py [hunk 5]
```diff
@@ -395,15 +351,15 @@ def transform_hamiltonian(h, generators, paulix_ops, paulix_sector):
     >>> geometry = np.array([[0.0, 0.0, -0.69440367], [0.0, 0.0, 0.69440367]])
     >>> mol = qml.hf.Molecule(symbols, geometry)
     >>> H = qml.hf.generate_hamiltonian(mol)(geometry)
-    >>> generators, paulix_ops = qml.hf.generate_symmetries(H, len(H.wires))
+    >>> generators, paulixops = qml.hf.generate_symmetries(H, len(H.wires))
```
Comment (obliviateandsurrender) [inline pennylane/hf/tapering.py:354]: Shouldn't this be removed?

File: pennylane/hf/tapering.py [hunk 6]
```diff
@@ -526,9 +482,9 @@ def transform_hf(generators, paulix_ops, paulix_sector, num_electrons, num_wires
     >>> mol = hf.Molecule(symbols, geometry, charge=1)
     >>> H = hf.generate_hamiltonian(mol)(geometry)
     >>> n_qubits, n_elec = len(H.wires), mol.n_electrons
-    >>> generators, paulix_ops = hf.generate_symmetries(H, n_qubits)
+    >>> generators, paulixops = hf.generate_symmetries(H, n_qubits)
```
Comment (obliviateandsurrender) [inline pennylane/hf/tapering.py:485]: Should be removed?

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]
Thanks @soranjh! Leaving a couple of small changes after which it should be good to go!

File: pennylane/hf/tapering.py [hunk 7]
```diff
@@ -468,7 +426,7 @@ def optimal_sector(qubit_op, generators, active_electrons):
     >>> geometry = np.array([[0., 0., -0.66140414], [0., 0., 0.66140414]])
     >>> mol = qml.hf.Molecule(symbols, geometry)
     >>> H = qml.hf.generate_hamiltonian(mol)(geometry)
-    >>> generators, paulix_ops = qml.hf.generate_symmetries(H, len(H.wires))
+    >>> generators, paulixops = qml.hf.generate_symmetries(H, len(H.wires))
```
Comment (obliviateandsurrender) [inline pennylane/hf/tapering.py:429]:
```suggestion
    >>> generators, paulixops = qml.hf.symmetry_generators(H)
```

File: tests/hf/test_tapering_hf.py [hunk 8]
```diff
@@ -312,23 +288,19 @@ def test_generate_paulis(generators, num_qubits, result):
                 qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(2)]),
                 qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(3)]),
             ],
-            [qml.PauliX(wires=[1]), qml.PauliX(wires=[2]), qml.PauliX(wires=[3])],
         ),
     ],
 )
-def test_generate_symmetries(symbols, geometry, num_qubits, res_generators, res_pauli_ops):
-    r"""Test that generate_symmetries returns the correct result."""
+def test_generate_symmetries(symbols, geometry, num_qubits, res_generators):
```
Comment (obliviateandsurrender) [inline tests/hf/test_tapering_hf.py:294]: Maybe change the test name to `test_symmetry_generators`

File: tests/hf/test_tapering_hf.py [hunk 9]
```diff
@@ -412,11 +384,11 @@ def test_cliford(generator, paulix_ops, result):
         ),
     ],
 )
-def test_transform_hamiltonian(symbols, geometry, generator, paulix_ops, paulix_sector, ham_ref):
+def test_transform_hamiltonian(symbols, geometry, generator, paulixops, paulix_sector, ham_ref):
     r"""Test that transform_hamiltonian returns the correct hamiltonian."""
     mol = qml.hf.Molecule(symbols, geometry)
     h = qml.hf.generate_hamiltonian(mol)()
-    ham_calc = transform_hamiltonian(h, generator, paulix_ops, paulix_sector)
+    ham_calc = qml.taper(h, generator, paulixops, paulix_sector)
```
Comment (obliviateandsurrender) [inline tests/hf/test_tapering_hf.py:391]: Should we use the imported function (`taper`) directly or the `qml.name`?

File: tests/hf/test_tapering_hf.py [hunk 10]
```diff
@@ -625,12 +597,13 @@ def test_hf_energy(symbols, geometry, charge):
...
-    generators, paulix_ops = generate_symmetries(hamiltonian, len(hamiltonian.wires))
+    generators = qml.symmetry_generators(hamiltonian)
+    paulixops = qml.paulix_ops(generators, len(hamiltonian.wires))
     paulix_sector = optimal_sector(hamiltonian, generators, mol.n_electrons)
 
-    hamiltonian_tapered = transform_hamiltonian(hamiltonian, generators, paulix_ops, paulix_sector)
-    hf_state_tapered = transform_hf(
-        generators, paulix_ops, paulix_sector, mol.n_electrons, len(hamiltonian.wires)
+    hamiltonian_tapered = qml.taper(hamiltonian, generators, paulixops, paulix_sector)
+    hf_state_tapered = taper_hf(
```
Comment (obliviateandsurrender) [inline tests/hf/test_tapering_hf.py:605]: Same as above.

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]

File: tests/hf/test_tapering_hf.py (hunk 9 above)
Comment (soranjh) [reply tests/hf/test_tapering_hf.py:391]: I prefer to use `qml.name` to make sure that these functions remain high-level and the user interface is always `qml.name`.

File: tests/hf/test_tapering_hf.py (hunk 10 above)
Comment (soranjh) [reply tests/hf/test_tapering_hf.py:605]: I prefer to use `qml.name` to make sure that these functions remain high-level and the user interface is always `qml.name`.

Comment (obliviateandsurrender) [reply tests/hf/test_tapering_hf.py:602]: Maybe then use `qml.taper_hf` here to be consistent throughout? But then I also wonder if it beats the purpose of importing the functions in the first place?

**Review by obliviateandsurrender — APPROVED** [review]
Looks good to me. Just left one final comment.

File: tests/hf/test_tapering_hf.py (hunk 10 above)
Comment (soranjh) [reply tests/hf/test_tapering_hf.py:602]: The functions `taper`, `symmetry_generators`, `paulix_ops` are not qchem-specific and they will be accessible with `qml.name`. The rest of the functions will either stay in qchem or go to a future home dedicated to operator arithmetic. So, these three functions are the only ones that will be accessible with `qml.name`.

File: pennylane/__init__.py (hunk 1 above)
Comment (ixfoduap) [reply pennylane/__init__.py:36]: deeper :wink:

**Review by ixfoduap — APPROVED** [review]

**Pull request merged**